src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp

```diff
@@ -0,0 +1,535 @@
+// Copyright (C) 2018-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/common_optimizations/fuse_moe_experts.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "itt.hpp"
+#include "openvino/core/graph_util.hpp"
+#include "openvino/core/node.hpp"
+#include "openvino/op/add.hpp"
+#include "openvino/op/broadcast.hpp"
+#include "openvino/op/concat.hpp"
+#include "openvino/op/constant.hpp"
+#include "openvino/op/convert.hpp"
+#include "openvino/op/divide.hpp"
+#include "openvino/op/gather.hpp"
+#include "openvino/op/matmul.hpp"
+#include "openvino/op/multiply.hpp"
+#include "openvino/op/non_zero.hpp"
+#include "openvino/op/one_hot.hpp"
+#include "openvino/op/reduce_sum.hpp"
+#include "openvino/op/reshape.hpp"
+#include "openvino/op/scatter_elements_update.hpp"
+#include "openvino/op/shape_of.hpp"
+#include "openvino/op/slice.hpp"
+#include "openvino/op/softmax.hpp"
+#include "openvino/op/split.hpp"
+#include "openvino/op/squeeze.hpp"
+#include "openvino/op/swish.hpp"
+#include "openvino/op/tile.hpp"
+#include "openvino/op/topk.hpp"
+#include "openvino/op/transpose.hpp"
+#include "openvino/op/unsqueeze.hpp"
+#include "openvino/op/util/shape_of_base.hpp"
+#include "openvino/pass/manager.hpp"
+#include "openvino/pass/pattern/matcher.hpp"
+#include "openvino/pass/pattern/op/block.hpp"
+#include "openvino/pass/pattern/op/pattern.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "transformations/rt_info/decompression.hpp"
+#include "transformations/utils/utils.hpp"
+namespace ov {
+namespace pass {
+
+using namespace ov::pass;
+using namespace ov::pass::pattern;
+
+namespace {
+
+// Check if a Slice operation's end parameter matches one found in model.
+// The pattern expects a 2-element constant [1, MAX] where.
+// This pattern matches Slice operations like: Slice(input, [0,0], [1,INT_MAX], [1,1], [0,1])
+// index_add__Slice_2 = makeOP<opset8::Slice>({index_add__ScatterElementsUpdate_5, {0,0}, {1,INT_MAX}, {1,1}, {0,1}});
+bool is_slice_to_end(const std::shared_ptr<Node>& node) {
+    auto constant = ov::as_type_ptr<ov::op::v0::Constant>(node);
+    if (!constant || !constant->get_element_type().is_integral_number()) {
+        return false;
+    }
+
+    const auto values = constant->cast_vector<int64_t>();
+    // Expect exactly 2 values: [1, MAX] where first dimension ends at 1
+    if (values.size() != 2 || values[0] != 1) {
+        return false;
+    }
+
+    // Second dimension should use maximum value to indicate slicing to the end
+    const auto end_value = values[1];
+    if (constant->get_element_type() == element::i32) {
+        return end_value == std::numeric_limits<int32_t>::max();
+    }
+
+    if (constant->get_element_type() == element::i64) {
+        return end_value == std::numeric_limits<int64_t>::max();
+    }
+
+    return false;
+}
+
+struct expert_data {
```

Comment:
I would prefer to define this structure in the header, then you can use it in the tests

Reply:
Moved to header